cdc/redo/common/redo.go

```diff
@@ -52,7 +57,54 @@ const (
 
 // LogMeta is used for store meta info.
 type LogMeta struct {
-	CheckPointTs   uint64           `msg:"checkPointTs"`
-	ResolvedTs     uint64           `msg:"resolvedTs"`
-	ResolvedTsList map[int64]uint64 `msg:"-"`
+	CheckPointTs uint64 `msg:"checkPointTs"`
+
+	ResolvedTsList map[int64]uint64 `msg:"resolvedTsList"`
+}
+
+// ResolvedTs returns the minimal resolved timestamp of all tables.
+// If no resolved timestamp is record, return checkpoint instead.
+func (z LogMeta) ResolvedTs() model.Ts {
+	var minResolvedTs uint64 = math.MaxUint64
+	for _, ts := range z.ResolvedTsList {
+		if ts < minResolvedTs {
+			minResolvedTs = ts
+		}
+	}
+	if minResolvedTs == math.MaxUint64 {
+		return z.CheckPointTs
+	}
+	return minResolvedTs
+}
+
+// MarshalMsg implements msgp.Marshaler
+func (z LogMeta) MarshalMsg(b []byte) ([]byte, error) {
+	buff := make([]byte, 16+16*len(z.ResolvedTsList))
+	binary.LittleEndian.PutUint64(buff[0:8], z.CheckPointTs)
+	binary.LittleEndian.PutUint64(buff[8:16], uint64(len(z.ResolvedTsList)))
+
+	tStart := 16
+	for tID, ts := range z.ResolvedTsList {
+		binary.LittleEndian.PutUint64(buff[tStart:tStart+8], uint64(tID))
+		binary.LittleEndian.PutUint64(buff[tStart+8:tStart+16], ts)
+		tStart += 16
+	}
+	b = append(b, buff...)
+	return b, nil
+}
+
+// UnmarshalMsg implements msgp.Unmarshaler
+func (z *LogMeta) UnmarshalMsg(b []byte) ([]byte, error) {
```

Comment:
Also curious about why the msgpack `MashalMsg` and `UnmarshalMsg` are replaced with these two customized functions?

Reply:
Seems msgp map only support string map key.